src/lib/Hydra/Controller/User.pm

```diff
@@ -154,6 +155,69 @@ sub google_login :Path('/google-login') Args(0) {
     doEmailLogin($self, $c, "google", $data->{email}, $data->{name} // undef);
 }
 
+sub github_login :Path('/github-login') Args(0) {
+    my ($self, $c) = @_;
+
+    error($c, "Logging in via GitHub is not enabled.") unless $c->config->{enable_github_login};
+    my $client_id = $c->config->{github_client_id} or die "github_client_id not configured.";
+    my $client_secret = $c->config->{github_client_secret} // do {
+        my $client_secret_file = $c->config->{github_client_secret_file} or die "github_client_secret nor github_client_secret_file is configured.";
+        my $client_secret = read_file($client_secret_file);
+        $client_secret =~ s/\s+//;
+        $client_secret;
+    };
+    die "No github secret configured" unless $client_secret;
+
+    my $ua = new LWP::UserAgent;
+    my $response = $ua->post(
+        'https://github.com/login/oauth/access_token',
+        {
+            client_id => $client_id,
+            client_secret => $client_secret,
+            code => ($c->req->params->{code} // die "No token."),
+        }, Accept => 'application/json');
+    error($c, "Did not get a response from GitHub.") unless $response->is_success;
+
+    my $data = decode_json($response->decoded_content) or die;
+    my $access_token = $data->{access_token} // die "No access_token in response from GitHub.";
+
+    $response = $ua->get('https://api.github.com/user/emails', Accept => 'application/vnd.github.v3+json', Authorization => "token $access_token");
+    error($c, "Did not get a response from GitHub for email info.") unless $response->is_success;
+
+    $data = decode_json($response->decoded_content) or die;
+    my $email;
+
+    foreach my $eml (@{$data}) {
+        $email = $eml->{email} if $eml->{verified} && $eml->{primary};
+        print STDERR "$eml->{email}\n";
+    }
+
+    print STDERR "$email\n";
+    $response = $ua->get('https://api.github.com/user', Authorization => "token $access_token");
+    error($c, "Did not get a response from GitHub for user info.") unless $response->is_success;
+    $data = decode_json($response->decoded_content) or die;
+
+    doEmailLogin($self, $c, "github", $email, $data->{name} // undef);
```

Comment:
Can `$email` be uninitialized here?

Reply:
I don't think github users can delete their primary email address, so this could only happen when github returns faulty data. I will add a `die` here to catch this though, as it seems like something useful to be caught